onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "model.h"
+#include "core/providers/nnapi/nnapi_builtin/builders/helper.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/nnapi_implementation.h"
+
+#ifdef USENNAPISHAREDMEM
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
+namespace onnxruntime {
+namespace nnapi {
+
+Model::Model() : nnapi_(NnApiImplementation()) {}
+
+Model::~Model() {
+  if (execution_)
+    nnapi_->ANeuralNetworksExecution_free(execution_);
+
+  nnapi_->ANeuralNetworksCompilation_free(compilation_);
+  nnapi_->ANeuralNetworksModel_free(model_);
+}
+
+void Model::AddInput(const std::string& name, const android::nn::wrapper::OperandType& operand_type) {
+  input_names_.push_back(name);
+  operand_types_.insert({name, operand_type});
+}
+
+void Model::AddOutput(const std::string& name, const android::nn::wrapper::OperandType& operand_type) {
+  output_names_.push_back(name);
+  operand_types_.insert({name, operand_type});
+}
+
+const std::vector<std::string>& Model::GetInputs() const {
+  return input_names_;
+}
+
+const std::vector<std::string>& Model::GetOutputs() const {
+  return output_names_;
+}
+
+const android::nn::wrapper::OperandType& Model::GetInputType(const std::string& name) const {
+  return operand_types_.at(name);
+}
+
+const android::nn::wrapper::OperandType Model::GetOutputType(const std::string& name) const {
+  const auto& output_type = operand_types_.at(name);
+  android::nn::wrapper::OperandType type(
+      output_type.type, shaper_for_exeuction_[name], output_type.operandType.scale, output_type.operandType.zeroPoint);
+
+  return type;
+}
+
+void Model::SetInputMap(std::unordered_map<std::string, size_t>&& input_map) {
+  input_map_ = std::move(input_map);
+}
+
+void Model::SetOutputMap(std::unordered_map<std::string, size_t>&& output_map) {
+  output_map_ = std::move(output_map);
+}
+
+size_t Model::GetMappedInputIdx(const std::string& name) const {
+  return input_map_.at(name);
+}
+
+size_t Model::GetMappedOutputIdx(const std::string& name) const {
+  return output_map_.at(name);
+}
+
+void Model::SetInputBuffer(const int32_t index, const InputOutputBuffer& input) {
+  if (!prepared_for_exe_) PrepareForExecution();
+  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_setInput(
```

Comment:
nit: put the body of the 'if' on the next line so you can set a breakpoint on it